Overview pages for themes and diagrams

johthor opened the pull request:
Create overview pages for all themes and all diagrams each as 

- Overview List
- Gallery View
- Table View
- List View

Create also dedicated pages per theme and per diagram showing each one in every possible variant of the other.

The-Lum commented:
Hello @johthor,

A good beginnig...
_You are an expert on Jekyll & Liquid..._

To avoid a lot of directories on the root directory...  Perhaps could you create 2 folders:
- `collections` or `jekyll_collections` or `site_collections` or ...
- `pages` or `site_pages` or `filter_by` or `by` or ...
>[![](https://www.planttext.com/api/plantuml/svg/SoWkIImgISlCIItcqa_Epyb9JIx9pCyhrezFoKnCBqhCBTP74o_9ICrD1OaM9AO3QIHb406Ol4DgNWhON000)](https://www.planttext.com/?text=SoWkIImgISlCIItcqa_Epyb9JIx9pCyhrezFoKnCBqhCBTP74o_9ICrD1OaM9AO3QIHb406Ol4DgNWhON000) 

_Good improvement,_ 👍 
Regards,
Th.

johthor commented:
Hi @The-Lum I think I have overextended the original scope of this PR by far 😁 

I have put in everything I could think of.

Unfortunately the original processing logic with two data files was no longer sufficient. I tried to change the scripts accordingly but I'm no Perl developer so I'm not sure if everything still works. Instead I wrote my own scripts.

I hope that we will be able to merge both approaches somehow. 

I fairly confident about the Jekyll related stuff. The rest not so much.

The-Lum commented:
Hi @johthor 

Thanks for this PR, _that makes a big enhancement on this gallerie..._

Before a review, here are some few questions or requests:
1. Could you put your `/script/input` on the initial folder on the root `/input`?
(Just to keep the history of this folder on git)
I Can change the scripts after accordingly.
1. What is the meaning of `asset` folder? Is it mandatory? Why old `../gallerie/img...` des not work?
1. What is your linter?
1. Then an open question about theme description file.... It must be interressing to automatise this... But some changes would be made directly on the plantuml repo, in order to have metadata and description with the themes. _[I Can take this action]_

Next step for me: a review, directly on the PR code.
Regards,
Th.

johthor commented:
Hi @The-Lum let me try to answer some of the questions.

1. There is no `/script/input/` directory. If you meant `/collections/_diagrams/input/` instead, moving the directory to the project root would no longer allow the diagram-specific pages to show the code used to create the diagrams. I use [Jekyll relative includes](https://jekyllrb.com/docs/includes/#including-files-relative-to-another-file) to show the static `puml` files which unfortunately does not allow for higher level directories as targets.  
As diagrams are no longer defined by just a PUML file but have become a full Jekyll collection and consist now of a `.md` description file and a corresponding `.puml` file I think that colocating the collection and its diagram definitions also makes sense.
2. The `assets` folder itself is part of the default Jekyll setup and where the [dinky theme](https://github.com/pages-themes/dinky/tree/master?tab=readme-ov-file#stylesheet) expects style related customizations.  
The other aspect is that I wanted to exclude the generated themed diagram definition files from the generated Site as they are only a temporary byproduct and not really necessary. But the `exclude` config only allowed to exclude the complete `/gallery` directory, so that the SVG in `/gallery/img` had to move somewhere else. And I first decided that `/assets/diagrams` would be fitting next to the generated `/assets/images`. But on closer inspection the diagram SVGs are not assets in a classical web design sense.  
So I'm open to moving them somewhere else as long as the intermediate themed diagram descriptions are excluded from the site.
3. I don't have a linter specifically but some GitHub action regularly complained about my work. So I tried to make it happy https://github.com/johthor/puml-themes-gallery/actions/runs/11636988084  
I'm not sure where exactly that action is triggered so I think it is more your linter 😁 
4. Regarding fully automating the theme description extraction. Yes it would have been nice but for now I used a simple heuristic that worked for almost all themes and produced good enough results. I'm not sure that it is worthwhile to fully automate the extraction as I don't believe many new themes to be created in the next month.  
None the less the code should still notice new themes and generate fallback collection files for that case. 

Regarding the intermediate themed diagram descriptions I might have an idea to get rid of them altogether and and still generating the same SVGs.

With themes and diagrams now being upgraded to Jekyll collections I have some more ideas for improved theme and diagram variant display but those shall be part of another issue.

The-Lum commented:
Hi @johthor,

Thanks for all those prescisions.

1. _Oups..._  I mix 2 questions
In fact question 1. was for `diagrams.csv` and `themes.csv` on ` \script`, but I understand now that Jekyll uses collection, those files is now used only by scripts on `/script`.
_[Private joke: we can now return to YAML file...]_

2. For Assets. I understand. 👍 

3. Good catch... That is my linter 🤣
_Historically, I put this GH Linter, and then I disable on my repo.; But it will be good to re-enable it on my repo..._

4. I can take the action to initiate a `%get_all_themes(full)`...

5. New question
>Regarding the intermediate themed diagram descriptions I might have an idea to get rid of them altogether and and still generating the same SVGs as PlantUML now supports a -theme option.

- What does the changed `generate_all_SVG.sh` script do?
Ah I understand, but my first goal with `!theme $theme` is to run only **once** `java -jar lib/plantuml.jar` (perf. & eco-conception, ♻️)


I think I will merge your PR _as it_.
_And then, I will make modification on scripts..._

Thanks for all your job. 🙏 
Regards,
Th.

The-Lum commented:
---

Just another question about:
- https://github.com/johthor/puml-themes-gallery/commit/1ee6a38f8f3a94cdd2ac9e902d3dec2e3a5f3eb1

What is the meaning of a list _(and definition lists)_ between:
```yml
Original Author: Author_Name
```
🆚 
```yml
Original Author
: Author_Name
```

---

Regards,
Th.

johthor commented:
Regarding `themes.csv|yml` and `diagrams.csv|yml`: Yes those files are no longer needed by Jekyll and so I moved them into `/script` which might not be the best place but as they are now only used by the scripts that was my first idea. In fact I think the `diagrams` file is now fully obsolete but I'm not sure about that.

Regarding the generation of themed PUML files: I did not take performance and eco into account. So how about generating the PUML files into `/gallery/input` or `/gallery/themed-input` and then moving back / converting the SVG files into `/gallery` and excluding the `/gallery/[themed-]input` directory from the site build. That might achieve both our goals.  
The `/gallery/input/*` files might even be removed from the git repository as well.

About the definition lists: Some of the themes (e.g. `mars`, `toy` or `vibrant`) provided more meta information besides the author and those looked to me like a definition list. _I don't really like the markdown syntax for those_
We could reduce the whitespace around the definition lists.

I will also break up the `run.sh` script and move some of it into `/script`. I don't think the collection creation functions will be of any use in the future.

The-Lum commented:
Hi @johthor,

_Once the machine starts it doesn't stop..._ 👍 Thanks for your job.

- Let me know when you think I can merge this PR?
And then you can propose another PR... ;-)

Regards,
Th.

johthor commented:
I really think, that now I have added everything I'm able to do now. There is still something strange with the table view but I'm not FrontEnd enough to fix that right now. ;)

So please merge as you want. I won't add more commits to this branch. _I promise ;)_

The-Lum commented:
Ok, thanks for the answer.

But sorry for new questions:
- You rechange the place of the generated SVG?
- Why `generate_all_SVG.sh` is deleted?
- What is the script that makes SVGs now?

I continue my review...

The-Lum commented:
Hi @johthor 

If you create `gallery/themed-input`:
- Why not to create also `gallery/SVG`

I have a little trouble with the unbalanced tree...
_But initially it was! Sorry to me..._

johthor commented:
No problem. I'm aware, that I might have changed to much to quick. ;(

* yes I reverted their place back to `/gallery` or at least similar to their former place. As they were not really assets and `/gallery` is a nicer name. And I solved my exclusion problem by switching `/gallery/*.puml` with `/gallery/themed-input/*puml` and `/gallery/img/*.svg` with `/gallery/*svg`
* I deleted `generate_all_SVG.sh` after you pointed out that running `plantuml` once would be more performant. So the original script should still be able to generate all diagram SVGs in one go. Sorry I did not communicate that or the reasons very well.  
**I just realized I made an error here** I wanted to roll the script back and must have deleted it instead. I'm really sorry about that and will fix that immediately.
* I think there is nothing wrong with `/gallery/SVG`. But might I suggest `/gallery/lightmode` or similar. _Hinting at yet another idea not yet converted into an issue ;)_

Guess not adding any more commits was a lie after all.

johthor commented:
Ah something went wrong. I couldn't test the real integration with GitHub Pages. I think the following actions should help

- Change `.ruby-version` to `3.3.4`
- Removing `baseUrl` and `url` from the `_config.yml` OR removing `url` and setting `baseUrl` to `/puml-themes-gallery`
- Maybe `figure.html` has also to be adjusted with some `{{ imageURL | relative_url }}` Liquid filter.

The-Lum commented:
Hi @johthor,

1/+2/: Done with https://github.com/The-Lum/puml-themes-gallery/commit/54d5433ff4bda295b6d05480032d7c9933231a57
And for 3/:
- What would be the explicitly change?

Regards,
Th.

johthor commented:
```
{%capture imageSrc %}/gallery/lightmode/{{ include.diagram.name }}-{{ include.theme.name}}.svg{% endcapture %}

<figure class="image-box{% if include.caption %} with-caption{% endif %}">
    <a href="{{ include.url | relative_url }}">
        <img src="{{ imageSrc | relative_url }}"
                alt="{{ include.diagram.display_name }} in {{ include.theme.display_name}}"
                loading="lazy" />
    </a>
{% if include.caption %}
    <figcaption>{{ include.caption }}</figcaption>
{% endif %}
</figure>
```

Which will at least fix some of the problems but there are more with every link to a page, which I don't understand and nothing did prepare me for this. I don't think, that spamming `{{ xyz | relative_url }}` should be the answer just because the page is not delivered from the root but from a subpath. 

I think there is still something missing but I won't find it today.

The-Lum commented:
Thanks a lot.

>I think there is still something missing but I won't find it today.

No problem.
That is not so urgent.

_Have a good night and week._
